cpp/src/hash/sha_hash.cu

```diff
@@ -0,0 +1,782 @@
+  // Digest size in bytes
+  static constexpr uint32_t digest_size = 128;
+  // Number of bytes used for the message length
+  static constexpr uint32_t message_length_size = 16;
+
+  void CUDA_DEVICE_CALLABLE hash_step(sha_intermediate_data& hash_state)
+  {
+    sha512_hash_step(hash_state);
+  }
+
+  sha_intermediate_data hash_state;
+};
+
+/**
+ * @brief Call a SHA-1 or SHA-2 hash function on a table view.
+ *
+ * @tparam Hasher The struct used for computing SHA hashes.
+ *
+ * @param input The input table.
+ * results.
+ * @param empty_result A string representing the expected result for empty inputs.
+ * @param stream CUDA stream on which memory may be allocated if the memory
+ * resource supports streams.
+ * @param mr Memory resource to use for the device memory allocation
+ * @return A new column with the computed hash function result.
+ */
+template <typename Hasher>
+std::unique_ptr<column> sha_hash(table_view const& input,
+                                 string_scalar const& empty_result,
+                                 rmm::cuda_stream_view stream,
+                                 rmm::mr::device_memory_resource* mr)
+{
+  if (input.num_columns() == 0 || input.num_rows() == 0) {
+    // Return the hash of a zero-length input.
+    return make_column_from_scalar(empty_result, input.num_rows(), stream, mr);
+  }
+
+  // Accepts string and fixed width columns.
+  // TODO: Accept single layer list columns holding those types.
+  CUDF_EXPECTS(
+    std::all_of(
+      input.begin(), input.end(), [](auto const& col) { return sha_type_check(col.type()); }),
+    "SHA unsupported column type");
+
+  // Result column allocation and creation
+  auto begin = thrust::make_constant_iterator(Hasher::digest_size);
+  auto offsets_column =
+    cudf::strings::detail::make_offsets_child_column(begin, begin + input.num_rows(), stream, mr);
+
+  auto chars_column =
+    strings::detail::create_chars_child_column(input.num_rows() * Hasher::digest_size, stream, mr);
+  auto chars_view = chars_column->mutable_view();
+  auto d_chars    = chars_view.template data<char>();
+
+  rmm::device_buffer null_mask{0, stream, mr};
+
+  auto const device_input = table_device_view::create(input, stream);
+
+  // Hash each row, hashing each element sequentially left to right
+  thrust::for_each(rmm::exec_policy(stream),
+                   thrust::make_counting_iterator(0),
+                   thrust::make_counting_iterator(input.num_rows()),
+                   [d_chars, device_input = *device_input] __device__(auto row_index) {
+                     Hasher hasher = Hasher{};
+                     for (int col_index = 0; col_index < device_input.num_columns(); col_index++) {
+                       if (device_input.column(col_index).is_valid(row_index)) {
+                         cudf::type_dispatcher<dispatch_storage_type>(
+                           device_input.column(col_index).type(),
+                           hasher,
```

Comment:
I think part of the issue may be making the `Hasher` both the thing that carries the hash buffer state _and_ the object that get's type dispatched. I would use separate objects, where `hasher` is passed as state to the dispatch function object.

```
template <typename Hasher>
struct dispatch_to_hash{
   Hasher * hash;
   column_device_view c;
   template <typename Element>
   auto operator()(size_type element_index){
      hash->process(c.element<Element>(element_index)):
   }
};
Hash hash{};
auto dispatch = dispatch_to_hash{&hasher, c};
...
cudf::type_dispatcher(..., dispatch);
```